src/vm/eventpipeblock.cpp

```diff
@@ -0,0 +1,129 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "common.h"
+#include "eventpipeblock.h"
+#include "eventpipeeventinstance.h"
+#include "fastserializableobject.h"
+#include "fastserializer.h"
+
+#ifdef FEATURE_PERFTRACING
+
+EventPipeBlock::EventPipeBlock(unsigned int maxBlockSize)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_NOTRIGGER;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    m_pBlock = new BYTE[maxBlockSize];
+    memset(m_pBlock, 0, maxBlockSize);
+    m_pWritePointer = m_pBlock;
+    m_pEndOfTheBuffer = m_pBlock + maxBlockSize;
+}
+
+EventPipeBlock::~EventPipeBlock()
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    if(m_pBlock != NULL)
+    {
+        m_pEndOfTheBuffer = NULL;
+        m_pWritePointer = NULL;
+        delete[] m_pBlock;
+        m_pBlock = NULL;
+    }
+}
+
+bool EventPipeBlock::WriteEvent(EventPipeEventInstance &instance)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    unsigned int totalSize = instance.GetAlignedTotalSize();
+    if (m_pWritePointer + totalSize >= m_pEndOfTheBuffer)
+    {
+        return false;
+    }
+
+    BYTE* alignedEnd = m_pWritePointer + totalSize + sizeof(totalSize); 
+
+    memcpy(m_pWritePointer, &totalSize, sizeof(totalSize));
+    m_pWritePointer += sizeof(totalSize);
+
+    unsigned int metadataId = instance.GetMetadataId();
+    memcpy(m_pWritePointer, &metadataId, sizeof(metadataId));
+    m_pWritePointer += sizeof(metadataId);
+
+    DWORD threadId = instance.GetThreadId();
+    memcpy(m_pWritePointer, &threadId, sizeof(threadId));
+    m_pWritePointer += sizeof(threadId);
+
+    LARGE_INTEGER timpeStamp = instance.GetTimeStamp();
+    memcpy(m_pWritePointer, &timpeStamp, sizeof(timpeStamp));
+    m_pWritePointer += sizeof(timpeStamp);
+
+    GUID activityId = instance.GetActivityId();
+    memcpy(m_pWritePointer, &activityId, sizeof(activityId));
+    m_pWritePointer += sizeof(activityId);
+
+    GUID relatedActivityId = instance.GetRelatedActivityId();
+    memcpy(m_pWritePointer, &relatedActivityId, sizeof(relatedActivityId));
+    m_pWritePointer += sizeof(relatedActivityId);
+
+    unsigned int dataLength = instance.GetDataLength();
```

Comment:
Anything that is associated with a length of ONE event should just be a 32 bit number (it is OK that we can't have 4GB sized events).     We will have to have the concept of IntPtr because EventSource already has that concept or I would argue to try to get rid of it there, but we should minimize its use as it DEFINATELY causes bugs (things that work on one bitness but not the other).

Reply:
I should also mention that ETW has a hard limit of 64K for the payload of its data (it uses a USHORT to represent it).   Thus as a practical matter all existing events are limited to about 64K for the foreseeable future.    (We can assert things at strategic places and have runtime checks at others if you like).

Reply:
That sounds fine to me.  We can leave it as is.  We don't need > 4GB events...